File: 2300.咒语和药水的成功对数.py
#
# @lc app=leetcode.cn id=2300 lang=python3
#
# [2300] 咒语和药水的成功对数
#

# @lc code=start
class Solution:
    def find_first_potion(self, nums: List[int], target: float) -> int:
        left=-1
        right=len(nums)
        while left+1 < right:
            mid = (left+right) // 2
            if nums[mid] >= target : #维护left的左边一直比target大
                left = mid
            else :
                right = mid
        return right
    def successfulPairs(self, spells: List[int], potions: List[int], success: int) -> List[int]:
        '''
        遍历sells数组，对于每一个咒语，算出所需的最低药水能量
        对potions降序排序，二分查找该最低药水能量的索引
        '''
        potions.sort(reverse=True)
        pairs = []
        for x in spells:
            target = success /x #python中浮点数和整数比较 回自动把整数转化为浮点数
            index = self.find_first_potion(potions, target)
            pairs.append(index)
        return pairs
# @lc code=end

File: 2300.咒语和药水的成功对数.cpp
/*
 * @lc app=leetcode.cn id=2300 lang=cpp
 *
 * [2300] 咒语和药水的成功对数
 */

// @lc code=start
class Solution {
public:
    int find_first_index(vector<int>& nums, long long target) {
        int left = -1, right = nums.size();
        while (left+1 < right) {
            int mid = left + (right - left) / 2;
            if (nums[mid] >= target) {
                left = mid;
            }
            else {
                right = mid;
            }
        }
        return right;
    }
    vector<int> successfulPairs(vector<int>& spells, vector<int>& potions, long long success) {
        sort(potions.begin(), potions.end(), [](int a, int b)
             { return a > b; });
        vector<int> pairs;
        for (int x : spells) {
            long long target = success % x ? success / x + 1 : success / x; //因为是向下取整 当非整除时加1
            int index = find_first_index(potions, target);
            pairs.push_back(index);
        }
        return pairs;
    }
};
// @lc code=end

